### PR TITLE
Multideterminant Batch code [Part 2 of Offload]

### PR DESCRIPTION
**Pull request by anbenali**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes
Batched Multideterminant implementation of the code. 
This PR belongs to the general effort aiming at generating a GPU implementation of the multideterminant code. #3830

This is part 2: All data needed for heavy kernels is now generated in batched code.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Refactoring (some api changes)
- Other (please describe):

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
- Laptop
- JLSE A100
- 
## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/Numerics/DeterminantOperators.h
Comment (markdewing) [inline src/Numerics/DeterminantOperators.h:243]: The return value list (curRatio_list) should be at the end of the argument list.   The general pattern for these functions seems to be input arguments first, then output arguments.

**Comment by prckent** [comment]
https://github.com/QMCPACK/qmcpack/pull/2926 was part 1?

File: src/Numerics/DeterminantOperators.h
Comment (prckent) [reply src/Numerics/DeterminantOperators.h:243]: Agreed. Thanks for the function description though.

File: src/Numerics/DeterminantOperators.h
Comment (ye-luo) [inline src/Numerics/DeterminantOperators.h:247]: Template on the value type instead of MatA and VecB. As a result, the type agreement between MatA and VecB can be enforced.
Replace curRatio_list RefVector with std::vector, there is no need of using reference_wrapper over scalars.

File: src/Numerics/DeterminantOperators.h
Comment (ye-luo) [inline src/Numerics/DeterminantOperators.h:330]: Remove the above mw_InverseUpdateByColumn and keep only this one. Also change template typenames to a value type shared by both matrices and vectors.

File: src/Numerics/DeterminantOperators.h
Comment (ye-luo) [inline src/Numerics/DeterminantOperators.h:333]: Let us keep `pragma omp parallel for` out.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp:67]: Remove restrict

File: src/Numerics/DeterminantOperators.h
Comment (ye-luo) [reply src/Numerics/DeterminantOperators.h:330]: RefVector -> const RefVector

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp:155]: Change this to std::vector then det0_list is not needed at all.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp:289]: Do we need this dummy function to compile the code? if not, remove.
If the single walker one is also needed, remove it as well.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h:328]: No documentation around both single and multi-walker functions. It is worth adding some documentation.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h:349]: This function has the same name as the one below for value only. I think it is better to rename the function with `_Grad` to improve readability. Please update both single and multi-walker versions.

File: src/Numerics/DeterminantOperators.h
Comment (anbenali) [reply src/Numerics/DeterminantOperators.h:31]: done

File: src/Numerics/DeterminantOperators.h
Comment (anbenali) [reply src/Numerics/DeterminantOperators.h:247]: Done

File: src/Numerics/DeterminantOperators.h
Comment (anbenali) [reply src/Numerics/DeterminantOperators.h:330]: Done

File: src/Numerics/DeterminantOperators.h
Comment (anbenali) [reply src/Numerics/DeterminantOperators.h:333]: done

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp
Comment (anbenali) [reply src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp:154]: I am not sure I understand how to do this. Ratio_list is already a ValueVector. I cannot use std::vector on it...

**Comment by anbenali** [comment]
Thanks for the fix @ye-luo 
Are we ready for this  one?

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp:66]: `const std::vector<ValueType> &`

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h:308]: params description is not in sync with the function.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp:569]: Remove the above two lines. Single walker function has removed them.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp:764]: Remove this line as the single walker function.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — DISMISSED** [review]
LGTM

**Review by ye-luo — APPROVED** [review]
LGTM need non-ANL approval

**Review by prckent — APPROVED** [review]
I noticed the sad "Finish this at some point" in the old MultiDiracDeterminant.h and InverseUpdateByColumn_GRAD . Not a reason to hold this PR up, but next time this area is touched either the comment should be removed (because the code is completed) or an abort added.

**Pull request merged**
